specs/interop/superchain_backend.md

```diff
@@ -0,0 +1,205 @@
+# Superchain Backend
+
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+**Table of Contents**
+
+- [Message Safety](#message-safety)
+- [Message Safety Resolution](#message-safety-resolution)
+  - [Unsafe to Cross-Unsafe](#unsafe-to-cross-unsafe)
+  - [Cross-Unsafe to (Safe|Finalized)](#cross-unsafe-to-safefinalized)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+The superchain backend provides the methods available for both the block builder and
+verifier for message safety checks. The backend can consumed over a JSON-RPC server,
+with the methods exposed under the `superchain` namespace, or as a go libary.
+
+```go
+type MessageSafetyLabel int
+
+const (
+    Invalid MessageSafetyLabel = iota
+    Unsafe
+    CrossUnsafe
+    Safe
+    Finalized
+)
+
+type SuperchainBackend interface {
+    MessageSafety(id Identifier, payload []byte) (MessageSafetyLabel, error)
+}
+```
+
+The backend maintains a view against the local network & registered peers in the [dependency set](./dependency_set.md).
+
+## Message Safety
+
+This method MUST enforce the [Identifier](./messaging.md#message-identifier)'s chainId is present in the chain's
+[dependency set](./dependency_set.md).
+
+This method MUST enforce the attributes listed in the [Identifier](./messaging.md#message-identifier) matches the log
+and block attributes returned by peer's `eth_getLogs` request with the `blockNumber` and `logIndex` parameters.
+
+This method MUST enforce the provided payload matches the bytes computed when
+constructing the [MessagePayload](./messaging.md#message-payload) against the fetched log.
+
+This method MUST return the applicable [safety label](./verifier.md#safety) for the message, or `Invalid` upon failures.
+
+This method MUST NOT enforce the [timestamp invariant](./messaging.md#timestamp-invariant) on the provided messages. The
```

Comment:
I am slightly confused by this - the timestamp invariant asserts that the message must not come from the future. We probably still want to assert this but we don't want to assert anything stricter than that (aka the policy)

Reply:
The reason we don't enforce it here is because we don't exactly know the exact context of the executing message.

For the block builder -> Timestamp is known when locally building or sealing the block
For the verifier -> Timestamp is known ahead of time

We could opt to pass the timestamp of the executing message as an argument in the `MessageSafety` but I was suggesting here that both the block builder / verifier locally enforce the invariant as the timestamp is apart of the message id and this api will ensure it's validity.

Reply:
If we have the blocknumber then we can fetch the timestamp by getting the block by number right? I also believe that both L1 and L2 timestamps are deterministic, so its possible we could have a cache of "starting timestamp" (which i believe exists already on the output oracle) and compute timestamps that way

Reply:
ah another example as well is the mempool validation. We wouldn't know the executing timestamp ahead of time

Reply:
changed this to "SHOULD NOT" instead and added a bit of comment amount why